src/Features/CSharp/Portable/EditAndContinue/SyntaxComparer.cs

```diff
@@ -2,213 +2,1555 @@
+                Debug.Assert(!TryGetParent(newRoot, out var _));
+                Debug.Assert(oldRootChildren.All(node => TryGetParent(node, out var parent) && parent == oldRoot));
+                Debug.Assert(newRootChildren.All(node => TryGetParent(node, out var parent) && parent == newRoot));
+            }
+        }
+
+        #region Tree Traversal
+
+        protected internal override bool TryGetParent(SyntaxNode node, [NotNullWhen(true)] out SyntaxNode? parent)
+        {
+            if (node == _oldRoot || node == _newRoot)
+            {
+                parent = null;
+                return false;
+            }
+
+            parent = node.Parent;
+            while (parent != null && !HasLabel(parent))
+            {
+                parent = parent.Parent;
+            }
+
+            return parent != null;
+        }
+
+        protected internal override IEnumerable<SyntaxNode>? GetChildren(SyntaxNode node)
+        {
+            Debug.Assert(HasLabel(node));
+
+            if (node == _oldRoot)
+            {
+                return _oldRootChildren;
+            }
+
+            if (node == _newRoot)
+            {
+                return _newRootChildren;
+            }
+
+            return HasChildren(node) ? EnumerateChildren(node) : null;
+        }
+
+        private IEnumerable<SyntaxNode> EnumerateChildren(SyntaxNode node)
+        {
+            foreach (var child in node.ChildNodes())
+            {
+                if (LambdaUtilities.IsLambdaBodyStatementOrExpression(child))
+                {
+                    continue;
+                }
+
+                if (HasLabel(child))
+                {
+                    yield return child;
+                }
+                else if (_compareStatementSyntax)
+                {
+                    foreach (var descendant in child.DescendantNodes(DescendIntoChildren))
+                    {
+                        if (HasLabel(descendant))
+                        {
+                            yield return descendant;
+                        }
+                    }
+                }
+            }
+        }
+        private bool DescendIntoChildren(SyntaxNode node)
+         => !LambdaUtilities.IsLambdaBodyStatementOrExpression(node) && !HasLabel(node);
 
-        protected const double ExactMatchDist = 0.0;
-        protected const double EpsilonDist = 0.00001;
+        protected internal sealed override IEnumerable<SyntaxNode> GetDescendants(SyntaxNode node)
+        {
+            var rootChildren = (node == _oldRoot) ? _oldRootChildren : (node == _newRoot) ? _newRootChildren : null;
+            return (rootChildren != null) ? EnumerateDescendants(rootChildren) : EnumerateDescendants(node);
+        }
+
+        private IEnumerable<SyntaxNode> EnumerateDescendants(IEnumerable<SyntaxNode> nodes)
+        {
+            foreach (var node in nodes)
+            {
+                if (HasLabel(node))
```

Comment:
>if (HasLabel(node)) [](start = 16, length = 19)

This shouldn't be necessary since we asserted in the constructor that all explicitly specified children have a label.

Reply:
This is still needed, because that assertion doesn't apply for statement syntax. So for example a simple field declaration, `int x = 1`, will analyze the EqualsValueExpression in case it has a lambda, so the child in that case is just a `NumericLiteralExpression`.